File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/ControlDrawerData.java
package net.kdt.pojavlaunch.customcontrols;

import static net.kdt.pojavlaunch.customcontrols.ControlDrawerData.Orientation.DOWN;
import static net.kdt.pojavlaunch.customcontrols.ControlDrawerData.Orientation.FREE;
import static net.kdt.pojavlaunch.customcontrols.ControlDrawerData.Orientation.LEFT;
import static net.kdt.pojavlaunch.customcontrols.ControlDrawerData.Orientation.RIGHT;
import static net.kdt.pojavlaunch.customcontrols.ControlDrawerData.Orientation.UP;

import android.content.Context;

import com.arata.yukarilauncher.R;
import com.arata.yukarilauncher.context.ContextExecutor;

import net.kdt.pojavlaunch.Tools;

import java.util.ArrayList;

@androidx.annotation.Keep
public class ControlDrawerData {

    public final ArrayList<ControlData> buttonProperties;
    public final ControlData properties;
    public Orientation orientation;

    @androidx.annotation.Keep
    public enum Orientation {
        DOWN,
        LEFT,
        UP,
        RIGHT,
        FREE
    }
/**
 * 「Orientations」の値を取得します。
 */
    public static String[] getOrientations(Context context) {
        return new String[]{
                context.getString(R.string.controls_orientation_down),
                context.getString(R.string.controls_orientation_left),
                context.getString(R.string.controls_orientation_up),
                context.getString(R.string.controls_orientation_right),
                context.getString(R.string.controls_orientation_free)
        };
    }
/**
 * 「orientation To Int」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public static int orientationToInt(Orientation orientation){
        switch (orientation){
            case DOWN: return 0;
            case LEFT: return 1;
            case UP: return 2;
            case RIGHT: return 3;
            case FREE: return 4;
        }
        return -1;
    }
/**
 * 「int To Orientation」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public static Orientation intToOrientation(int by){
        switch (by){
            case 0: return DOWN;
            case 1: return LEFT;
            case 2: return UP;
            case 3: return RIGHT;
            case 4: return FREE;
        }
        return null;
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public ControlDrawerData(){
        this(new ArrayList<>());
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public ControlDrawerData(ArrayList<ControlData> buttonProperties){
        this(buttonProperties, new ControlData(ContextExecutor.getString(R.string.controls_add_control_drawer), new int[] {}, Tools.currentDisplayMetrics.widthPixels/2f, Tools.currentDisplayMetrics.heightPixels/2f));
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public ControlDrawerData(ArrayList<ControlData> buttonProperties, ControlData properties){
        this(buttonProperties, properties, Orientation.LEFT);
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public ControlDrawerData(ArrayList<ControlData> buttonProperties, ControlData properties, Orientation orientation){
        this.buttonProperties = buttonProperties;
        this.properties = properties;
        this.orientation = orientation;
    }
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public ControlDrawerData(ControlDrawerData drawerData){
        buttonProperties = new ArrayList<>(drawerData.buttonProperties.size());
        for(ControlData controlData : drawerData.buttonProperties){
            buttonProperties.add(new ControlData(controlData));
        }
        properties = new ControlData(drawerData.properties);
        orientation = drawerData.orientation;
    }

}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/ControlDrawerData.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols

import android.content.Context
import com.arata.yukarilauncher.R
import com.arata.yukarilauncher.context.ContextExecutor
import com.arata.yukarilauncher.Tools
import java.util.ArrayList

/**
 * コントロールドロワーの設定データを保持するクラス。
 * ドロワーは折りたたみ可能なボタンコンテナで、複数のサブボタンを格納できます。
 */
@androidx.annotation.Keep
class ControlDrawerData {
    /** ドロワーの開閉方向を表す列挙型。 */
    @androidx.annotation.Keep
    enum class Orientation {
        /** 下方向に開く */
        DOWN,
        /** 左方向に開く */
        LEFT,
        /** 上方向に開く */
        UP,
        /** 右方向に開く */
        RIGHT,
        /** 自由位置に配置 */
        FREE
    }

    /** ドロワー内のサブボタンデータリスト。 */
    val buttonProperties: ArrayList<ControlData>
    /** ドロワー自体のプロパティ。 */
    val properties: ControlData
    /** ドロワーの開閉方向。 */
    @androidx.annotation.Keep
    var orientation: Orientation = Orientation.LEFT

    constructor() : this(ArrayList())

    constructor(buttonProperties: ArrayList<ControlData>) : this(buttonProperties, ControlData(ContextExecutor.getString(R.string.controls_add_control_drawer), intArrayOf(), Tools.currentDisplayMetrics.widthPixels / 2f, Tools.currentDisplayMetrics.heightPixels / 2f))

    constructor(buttonProperties: ArrayList<ControlData>, properties: ControlData) : this(buttonProperties, properties, Orientation.LEFT)

    constructor(buttonProperties: ArrayList<ControlData>, properties: ControlData, orientation: Orientation) {
        this.buttonProperties = buttonProperties
        this.properties = properties
        this.orientation = orientation
    }

    /** @param drawerData コピー元のControlDrawerData */
    constructor(drawerData: ControlDrawerData) {
        buttonProperties = ArrayList(drawerData.buttonProperties.size)
        for (controlData in drawerData.buttonProperties) {
            buttonProperties.add(ControlData(controlData))
        }
        properties = ControlData(drawerData.properties)
        orientation = drawerData.orientation
    }

    companion object {
        /**
         * ローカライズされた方向名の配列を返します。
         * @param context Androidコンテキスト
         * @return 方向名の文字列配列
         */
        fun getOrientations(context: Context): Array<String> {
            return arrayOf(
                context.getString(R.string.controls_orientation_down),
                context.getString(R.string.controls_orientation_left),
                context.getString(R.string.controls_orientation_up),
                context.getString(R.string.controls_orientation_right),
                context.getString(R.string.controls_orientation_free)
            )
        }

        /**
         * Orientation列挙型を整数値に変換します。
         * @param orientation 変換する方向
         * @return 整数値 (0=DOWN, 1=LEFT, 2=UP, 3=RIGHT, 4=FREE)
         */
        fun orientationToInt(orientation: Orientation): Int {
            return when (orientation) {
                Orientation.DOWN -> 0
                Orientation.LEFT -> 1
                Orientation.UP -> 2
                Orientation.RIGHT -> 3
                Orientation.FREE -> 4
            }
        }

        /**
         * 整数値をOrientation列挙型に変換します。
         * @param by 変換する整数値
         * @return 対応するOrientation、該当なしの場合はnull
         */
        fun intToOrientation(by: Int): Orientation? {
            return when (by) {
                0 -> Orientation.DOWN
                1 -> Orientation.LEFT
                2 -> Orientation.UP
                3 -> Orientation.RIGHT
                4 -> Orientation.FREE
                else -> null
            }
        }
    }
}
